pkg/jobs/job_info_storage.go

```diff
@@ -13,18 +13,174 @@ package jobs
 import (
 	"bytes"
 	"context"
+	"math"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 )
 
+// ProgressStorage reads and writes progress rows.
+type ProgressStorage jobspb.JobID
+
+// ProgressStorage returns a new ProgressStorage with the passed in job and txn.
+func (j *Job) ProgressStorage() ProgressStorage {
+	return ProgressStorage(j.id)
+}
+
+// Get returns the latest progress report for the job along with when it was
+// written. If the fraction is null it is returned as NaN, and if the resolved
+// ts is null, it is empty.
+func (i ProgressStorage) Get(ctx context.Context, txn isql.Txn) (float64, hlc.Timestamp, time.Time, error) {
+	ctx, sp := tracing.ChildSpan(ctx, "get-job-progress")
+	defer sp.Finish()
+
+	// The sort and limit should not be required since Write deletes all prior
+	// rows before inserting so there should only be one row.
+	row, err := txn.QueryRowEx(
+		ctx, "job-progress-get", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		"SELECT written, fraction, resolved FROM system.job_progress WHERE job_id = $1 ORDER BY written DESC LIMIT 1", i,
+	)
+
+	if err != nil || row == nil {
+		return 0, hlc.Timestamp{}, time.Time{}, err
+	}
+
+	written, ok := row[0].(*tree.DTimestampTZ)
+	if !ok {
+		return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DTimestampTZ (was %T)", row[1])
+	}
+
+	var fraction float64
+	if row[1] == tree.DNull {
+		fraction = math.NaN()
+	} else {
+		fracDatum, ok := row[1].(*tree.DFloat)
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DFloat (was %T)", row[1])
+		}
+		fraction = float64(*fracDatum)
+	}
+
+	var ts hlc.Timestamp
+	if row[2] != tree.DNull {
+		resolved, ok := row[2].(*tree.DDecimal)
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DDecimal (was %T)", row[1])
+		}
+		ts, err = hlc.DecimalToHLC(&resolved.Decimal)
+		if err != nil {
+			return 0, hlc.Timestamp{}, time.Time{}, err
+		}
+	}
+
+	return fraction, ts, written.Time, nil
+}
+
+// Write writes a progress update. If fraction is NaN or resolved is empty, that
+// field is left null. The time at which the progress was reported is recorded.
+func (i ProgressStorage) Write(ctx context.Context, txn isql.Txn, fraction float64, resolved hlc.Timestamp) error {
+	ctx, sp := tracing.ChildSpan(ctx, "write-job-progress")
+	defer sp.Finish()
+
+	if _, err := txn.ExecEx(
```

Comment:
hrm, given how frequently jobs update progress, i wonder if our write _and_ read paths would be faster if we only issued a delete tombstone once every N updates.

Reply:
this is exactly what we do to the legacy_progress key in the info table and that works fine.